public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -1,234 +1,262 @@
-      }
-      seriesList.sort(sortSeriesByLabel);
-      seriesList = this.transformToHistogramOverTime(seriesList);
-      return seriesList;
-    } else if (prometheusResult) {
-      const seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        if (response.data.data.resultType === 'matrix') {
-          seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-        } else if (response.data.data.resultType === 'vector') {
-          seriesList.push(this.transformInstantMetricData(metricData, options));
-        }
-      }
-      return seriesList;
-    }
+import {
+  ArrayVector,
+  DataFrame,
+  Field,
+  FieldType,
+  formatLabels,
+  TIME_SERIES_TIME_FIELD_NAME,
+  TIME_SERIES_VALUE_FIELD_NAME,
+} from '@grafana/data';
+import { FetchResponse } from '@grafana/runtime';
+import templateSrv from 'app/features/templating/template_srv';
+import {
+  PromDataSuccessResponse,
+  PromMatrixData,
+  PromMetric,
+  PromValue,
+  PromVectorData,
+  TransformOptions,
+} from './types';
+
+type MatrixOrVectorResult = PromMatrixData['result'][0] | PromVectorData['result'][0];
+
+export function transform(response: FetchResponse<PromDataSuccessResponse>, options: TransformOptions) {
+  const prometheusResult = response.data.data;
+
+  if (!prometheusResult.result) {
     return [];
   }
 
-  transformMetricData(metricData: any, options: any, start: number, end: number): TimeSeries {
-    const dps = [];
-    const { name, labels, title } = this.createLabelInfo(metricData.metric, options);
+  if (prometheusResult.resultType === 'scalar') {
+    return [
+      {
+        meta: options.meta,
+        refId: options.refId,
+        length: 1,
+        fields: [getTimeField([prometheusResult.result]), getValueField([prometheusResult.result])],
+      },
+    ];
+  }
 
-    const stepMs = parseFloat(options.step) * 1000;
-    let baseTimestamp = start * 1000;
+  if (options.format === 'table') {
+    const tableData = transformMetricDataToTable(prometheusResult.result, options);
+    return [tableData];
+  }
 
-    if (metricData.values === undefined) {
-      throw new Error('Prometheus heatmap error: data should be a time series');
-    }
+  const dataFrame: DataFrame[] = [];
+
+  prometheusResult.result.forEach((data: MatrixOrVectorResult) => dataFrame.push(transformToDataFrame(data, options)));
+
+  if (options.format === 'heatmap') {
+    dataFrame.sort(sortSeriesByLabel);
+    const seriesList = transformToHistogramOverTime(dataFrame);
+    return seriesList;
+  }
+
+  return dataFrame;
+}
+
+/**
+ * Transforms matrix and vector result from Prometheus to DataFrame
+ */
+function transformToDataFrame(data: MatrixOrVectorResult, options: TransformOptions): DataFrame {
+  const { name } = createLabelInfo(data.metric, options);
 
-    for (const value of metricData.values) {
+  const fields: Field[] = [];
+
+  if ('values' in data) {
```

Comment:
Could we change this to something like `if (isMatrixData(data))`? I think it would be clearer as I had to do 2 steps to check the typing to see that this is actually to check between metrics and vector type of result. If done as a type guard it should probably not be a problem for TS to keep working the same way.

Reply:
sure, good point.